Just testing whether this can be compiled on other platforms (multi instances)

anr2me opened the pull request.

anr2me committed:
Attempt to fix random crashing issue during Multiplayer.

anr2me committed:
Attempt to fix random crashing issue during Multiplayer.

anr2me committed:
* Detect multiple instance of PPSSPP running and automatically change the IP and MAC of the 2nd (or more) instance of PPSSPP to allow multiplayer on the same computer.
* Added Port-shifting support for games that use privileged ports (< 1024), all players need to have the same port offset.
  TODO: Add GUI for Port Offset to be configurable by players.

anr2me committed:
* Detect multiple instance of PPSSPP running and automatically change the IP and MAC of the 2nd (or more) instance of PPSSPP to allow multiplayer on the same computer.
* Added Port-shifting support for games that use privileged ports (< 1024), all players need to have the same port offset.
  TODO: Add GUI for Port Offset to be configurable by players.

anr2me committed:
Ooops forgot another "="

anr2me committed:
Ooops forgot another "="

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Fixed an issue where a peer considered as getting timed out right after it joined.

anr2me committed:
Fixed an issue where a peer considered as getting timed out right after it joined.

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Attempt to fix compilation errors for non-Windows system

anr2me committed:
Attempt to fix compilation errors for non-Windows system

sum2012 commented:
Seem compile ok

xiushudongfang commented:
Amazing,test some games,and it works fine.

hrydgard commented:
So do you actually want me to look at it for merging, or not? :) 

If so, please rebase it on master first, it doesn't need all those "merge" commits.

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Attempt to fix corrupted registers after returning from adhoc matching callbacks. (Fixed Dissidia 012 crashing or getting a blank screen when playing with more than 2 players, and probably some other games with similar issue)

anr2me committed:
Attempt to fix corrupted registers after returning from adhoc matching callbacks. (Fixed Dissidia 012 crashing or getting a blank screen when playing with more than 2 players, and probably some other games with similar issue)

ANR2ME commented:
I was just making sure it's up to date, and has no compilation issue.
Anyone could pick whichever commits i made and make a pull request from their own fork.

sum2012 commented:
I want test myself before pick up commit

lightfirst commented:
cool, multiple instance of PPSSPP support you add working fine 
Port-shifting support for games that use privileged ports (< 1024) also woring. ( I set portOffset = 5000 and tested on android)

sum2012 commented:
@lightfirst Can you help to test Metal Slug XX (ULUS10495) ? Thanks
Android host still fail (Client is windows 7)
I have set portOffset = 5000

lightfirst commented:
@sum2012  No problem for me. Have you change Core/HLE/proAdhoc.cpp
uint16_t portOffset = 0; //5000 to uint16_t portOffset = 5000;

LunaMoo committed:
Add an option for portOffset.

LunaMoo commented:
Maybe the port it got is already being used by something else on one of your devices?
 Well it needs an option anyway to keep compatibility with PSP/JPCSP, so might as well allow setting custom value, you can try https://github.com/LunaMoo/ppsspp/tree/adhoc_stuff where I just added that.

anr2me committed:
Merge branch 'adhoc_stuff' of https://github.com/LunaMoo/ppsspp

* Fixed minor Conflicts:
	Core/HLE/proAdhoc.cpp
* Changed default value of portOffset to 0 (to prevent people who don't know about it from asking why they can't play on coldbird's anymore)
* Changed the maximum portOffset value higher (afterall network port is in the range of 0-65535)

ANR2ME commented:
Thank for adding the port offset GUI @LunaMoo  :+1: 
I've merged it to my branch and also made minor modification to the default and max value.

@sum2012 and @lightfirst : Even when port-offset is set to 5000 it doesn't mean the actual port will be used is 5000, it's just an offset. If the game previously using port 80, after shifted it will be 5080 and if port 5080 is being used by other application, it won't work either, so you need to changed the offset to something else.

LunaMoo commented:
Cool, I forgot about Coldbird being default now, even through I probably looked at it while adding this option.:P

sum2012 commented:
@lightfirst  @LunaMoo 
Maybe re-compile problem.
First time I really forgot change portOffset 
Secord time I changed then compile.
Now I delete android\obj folder ,compile ,play ok

xiushudongfang commented:
How about this?
![qq 20151103114348](https://cloud.githubusercontent.com/assets/8541580/10900895/8e96f222-8220-11e5-9279-0d20db46a89a.png)
![qq 20151103114406](https://cloud.githubusercontent.com/assets/8541580/10900896/8e9b5a38-8220-11e5-8646-22d776eb9ca6.png)

ANR2ME commented:
LOL it doesn't really need a checkbox, afterall without port shifting it's the same as giving 0 port offset :) which is the value i used as default and noted as "(0 = PSP Compatibility)" in the option by LunaMoo

xiushudongfang commented:
Oh, I didn't know that LunaMoo had already do this before I replied.....

ANR2ME commented:
Btw, have anyone tried running multiple instance of PPSSPP on Linux and whether the automatic unique IP & MAC assignment for the 2nd, 3rd(, and so on) instances worked or not? 

I never tested it on linux and worried that the PPSSPP_ID will reset back to 1 everytime a new instance is started which could cause of them having the same IP (and mac)

lightfirst commented:
I will try linux later.

ANR2ME commented:
Btw @LunaMoo  are you sure that is how config value suppose to be used? 
isn't assigning the value directly at variable definition only worked for constant value? and what will happen if a player wanted to change the value in the middle of the game and then restart the rom, will portOffset variable be able to get the new value?

anr2me committed:
Making sure portOffset is being set when player changed the value after the game running/restarted.

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

LunaMoo commented:
Oh yeah sorry, was doing too many things at once yesterday, I suppose it wasn't even being set there from config variable at all ugh.~_~

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Separate Adhoc Matching Event's buffer for each event instead of sharing one buffer for all events since different event may need longer time than the other to be finished. (32bit version can play Dissidia 012 with 6 players now, while 64bit have 1 player crashed)

PS: It might not be the best approach to fix memory corruption during adhoc event callbacks, but it helped makes a bit more stable multiplayer (at least for 32bit version)

lightfirst commented:
compile PPSSPP-linux-32bit success after I copy libGl.so from /usr/lib/x86_64-linux-gnu to /usr/lib.
I am using 64bit ubuntu, how to compile 64bit PPSSPP on linux?

lightfirst commented:
running multiple instance of PPSSPP on Linux not working, the 2nd PPSSPP will close.
![p1](https://cloud.githubusercontent.com/assets/4612168/10911609/92aabc44-8280-11e5-8ae3-cd87c4511952.png)


portOffset variable not get the new value unless resatart game.

ANR2ME commented:
Thanks for testing @lightfirst :)
Since the log of AdhocServer showing "Already Existing IP" i guess PPSSPP_ID was reset back to 1 on every new instance instead of increasing it :(

sum2012 commented:
@lightfirst You need update commit https://github.com/ANR2MERefork/ppsspp/commit/e3727110f7a14bb0fa28afd49a1db5f80e2edae8
https://github.com/ANR2MERefork/ppsspp/commit/69bcc2ed3cad75e20a899cd356435f019f6c125b

@ANR2ME I think that now we can pull this commit into master ,do you agree ?

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Attempt to fix multiple instance of PPSSPP on Linux

ANR2ME commented:
@sum2012 There is still multiple instance issue on Linux, and compatibility issue between 32bit and 64bit version

@lightfirst I've made a new commit to fix multiple issue on Linux, could you give it a test again, Thanks.

lightfirst commented:
@sum2012 ok will try update
@ANR2ME LINUX/MACOSX have other issue with multipe. 
1st, all PPSSPPSDL will share  same save path(home/user/.config/ppsspp), that make PPSSPP share same ppsspp.ini and savedata. when play multipe, different should have it's own save paht
2nd, PPSSPPSDL using 2x windows size, I dont know how to setting it to be 1X windows size
@LunaMoo 0 to 60000 is to broad, 0 to 5000 is enough.

ANR2ME commented:
Hmm.. after testing 32bit vs 64bit on Dissidia 012 it seems Network sync failure issue was caused by the character instead of PPSSPP (ie. Vaan's Finishing move could cause Network sync failure)

@lightfirst 
0 to 60000 is normal and helped people to find the best port for them as different people have different applications running and there might be some that use port within 5000-ish (ie. YM, GTalk, etc)

I agree about using different config for each instance, but when i implement this my intention was to make debugging and testing multiplayer easier so giving unique IP and MAC was sufficient (at most is probably giving unique Username also, for games that differentiate players by Username)
Anyway, it's not like anyone could really enjoy playing multiplayer game using a single computer, especially action/real-time games (i don't enjoy games that isn't in fullscreen for example :D)

Btw, since you're saying that the 2nd instance in linux always have 2x Window size that means you can play multiplayer using multiple instance on Linux right?

sum2012 commented:
@ANR2ME Yes,I have't tested windows 32 bit Vs 64 bit

lightfirst commented:
Oh, I forget windows size can be change just using mouse.

I update core but running multiple instance of PPSSPP on Linux still not working

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

lightfirst commented:
Seem compile MACOSX ok
![ppssppsdl](https://cloud.githubusercontent.com/assets/4612168/10989466/7fca8592-8482-11e5-8584-8e80a7fda303.png)

ANR2ME commented:
does multiplayer using multiple instance works on OSX?

lightfirst commented:
@ANR2ME yes, I need use port offset > 1024 on linux/OSX or using sudo get root.

lightfirst commented:
multiplayer using multiple instance not works on OSX

sum2012 commented:
@ANR2ME Can we commit port offset stuff first ?

RowdyRocket commented:
This is awesome! Thanks for all your effort! Pretty cool playing four player Monster Hunter! I'm just having one problem. it only detects my first Xbox 360 controller. I have no problem detecting the others on JPCSP or any other program I've tried. Currently, as a work around, I'm using AntiMicro which limits me to two players because the keyboard window has to be active. Has anyone else had any luck detecting other x360 controllers besides the first? Or is there a way to have PPSSPP receive background keyboard input?

lightfirst commented:
XinputDevice.cpp limited to one Xinput device? #8250

RowdyRocket commented:
How would I go about downloading the Source of the modified files? I am working on the xbox controller issue and have made some progress. I'd like to try it with this modification.

RowdyRocket commented:
Sorry guys I'm new at this. How do I grab the source files for this modification? I'm playing with implementing the additional xbox controllers and have it functioning but would like to try it out with this version if I could. Thanks for the help.

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

unknownbrackets commented on Core/HLE/sceNet.cpp:
```diff
@@ -43,6 +50,168 @@ static struct SceNetMallocStat netMallocStat;
+
+	// Create shared memory object 
+
+	hIDMapFile = shm_open(ID_SHM_NAME, O_CREAT | O_RDWR, 0);
```
I believe this is supposed to start with a `/`.  Alternatively, we could use a file inside `$TMP`, I mean ultimately all we need is an instance number, right?

ANR2ME replied:
Yes, it just need a unique number per instance which will be used as the last byte of localhost IP. My plan was to increase the instance number on each additional instance, and reset the number when all instances has been closed/exited so the next first instance will be 1 again.

unknownbrackets replied:
But it could even just cycle around 256 even when you can't tell how many instances are alive right?  This would not give each instance a "stable" number, but at least they would have different ones.

-[Unknown]

ANR2ME replied:
Yes, and currently it's like that, since CloseHandle(hIDMapFile); / shm_unlink(ID_SHM_NAME); which is responsible to reset the number are being commented.

unknownbrackets commented on Core/HLE/sceKernelThread.cpp:
```diff
@@ -3095,6 +3100,13 @@ void __KernelExecuteMipsCallOnCurrentThread(u32 callId, bool reschedAfter)
 	call->savedId = cur->currentMipscallId;
 	call->reschedAfter = reschedAfter;
+	
+	// Also need to backup other regs which might be used by the called mips function, needed to prevent corrupting the original regs (ie. Dissidia 012 with more than 2 players, Lord of Arcana)
```
I don't like these changes.

First off, why is the callback being triggered at times when this would be a concern?  It's very unlikely this would be happening on the PSP.

If it does, you I think this should be handled inside `AfterMatchingMipsCall` instead of here, since I do not think this is the correct behavior for other mipscalls.

-[Unknown]

ANR2ME replied:
The thing is AfterMatchingMipsCall didn't always get executed, even within a few minutes timeout when i tested it (i used 250ms as matching callback timeout as it could cause slowdown/lag if i set too high, since most of the time it only takes less than 100ms), i believed the failing AfterMatchingMipsCall was caused by corrupted registers/memory.

unknownbrackets replied:
Well, that may mean that the callback called something else, or generated another callback.  Right now, callbacks inside other callbacks don't work.

Also, you probably need/want "real clock sync" during netplay.

-[Unknown]

ANR2ME replied:
I do agree about synced clock, but most games only use timestamp delta from previously recorded timestamp as networking timeout, so synced time will probably not affecting much to the stability.
However the scaled timestamp does affect network stability greatly during my test, especially when each instance is getting different/unstable FPS.

anr2me committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

sum2012 commented:
I have separate port offset 's pull request in #8496

lightfirst commented:
multiple instance of PPSSPP working on LINUX if using sudo
![5](https://cloud.githubusercontent.com/assets/4612168/19419610/d93bc3a2-940d-11e6-972e-92bf4b854eab.jpg)

hrydgard commented:
Alright, what do we do with this? Now that port offsets are merged, are there other important things here we need? 

Either way, it needs a rebase.

influxWishdream commented:
I think it's a good idea to rebase it now since its been 9 months since this last updated. Besides, the remaining code left provides fixes to the Linux and Android side of Port Offsetting as well as parts of networking that fixes problems on games like Phantasy Star Portable and Monster Hunter.

Could always start a new branch if there is something to fix.

ANR2ME commented:
Since Phantasy Star Portable 2 need to have Port Offset = 0 in order to work, there might be another network function that uses port but didn't shift the port number.

influxWishdream commented:
It's possible. I currently don't have a socket sniffer to check out which ports are used and what functions to look up. I'll try once power is back. In a middle of a typhoon at the moment.

hrydgard commented:
Either way it needs a rebase, @ANR2ME  are you up for it?

ANR2ME commented:
I think we should close this one instead, and move discussion to port offset commit that was merged for further improvement. Since this commit was only used to test whether it can be compiled on non-Windows platform :)
Was it this one? https://github.com/hrydgard/ppsspp/issues/8496

influxWishdream commented:
I'm not so sure, I think rebasing it is a better idea because recently the
port 0 to psp2 does not work on other builds than this one for some reason.
Could there have been a regression?

On Wed, Oct 19, 2016, 11:55 PM ANR2ME <notifications@github.com> wrote:

> I think we should close this one instead, and move discussion to port
> offset commit that was merged for further improvement. Since this commit
> was only used to test whether it can be compiled on non-Windows platform :)
> Was it this one? #8496 <https://github.com/hrydgard/ppsspp/pull/8496>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hrydgard/ppsspp/pull/8134#issuecomment-254856806>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEcd52B3BcUz1X4mEpkmHekxI0HE6tI_ks5q1j2CgaJpZM4GZrZ0>
> .
>

ANR2ME commented:
Hmm.. probably related to "callbacks inside other callbacks don't work" mentioned by @unknownbrackets above. Where this commit have a workaround to make games that use AdhocMatching a bit more stable (but not perfect)

influxWishdream commented:
Most likely would be the case. Either way, it might be a good idea to
include the workaround for now and rewrite it later. Maybe add a TODO or a
comment there that says so.

On Thu, Oct 20, 2016, 1:07 AM ANR2ME <notifications@github.com> wrote:

> Hmm.. probably related to "callbacks inside other callbacks don't work"
> mentioned by @unknownbrackets <https://github.com/unknownbrackets> above.
> Where this commit have a workaround to make games that use AdhocMatching a
> bit more stable (but not perfect)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hrydgard/ppsspp/pull/8134#issuecomment-254877259>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEcd58mLfWUZGZ2ctgdX_MrEksmvAV-Cks5q1k44gaJpZM4GZrZ0>
> .
>

sum2012 commented:
I think that it might be need rebased again when @adenovan done network chat in #7151

ANR2ME commented:
chat and callback workaround aren't related to port offset, so we should make a separated commits right? 
Anyway,  i will probably delete this old fork and reclone from official repo whenever i wanted to try something in the future. One of the flaw with port shifting is that all players need to have the same port offset, while different computer may have different port usage, one of the player may not have the port available, and not everyone familiar with ports, tunneling might be better since each player can use different tunnel port, and it's easier to forward since the port is known

influxWishdream commented:
Right, tunneling is so much better without screwing up the ports.
Well. At this point I'm not so sure.

I'll leave the decision up to you guys.

hrydgard commented:
How would we go about tunneling? make a little custom packet header containing the destination port, then simulating that when sending the packets into the sceNet libraries? or are there suitable ready made protocols we should use?

adenovan commented:
Hmm for tunneling can we use STTP https://en.m.wikipedia.org/wiki/Secure_Socket_Tunneling_Protocol and use Https port for the TLS it will pass many firewall I think and we need to think for people behind the nat without upnp enabled like android using cellular data. How can torrent software established p2p connection behind nat easily,  do we really need a server to do hole punching for user that not having upnp enabled like Skype do?

ANR2ME commented:
Well for performance reason Encryption isn't really needed as it will give extra overhead, but having encryption may improve security (ie. tampering game data for cheating).

Anyway, i was thinking a simple tunneling by wrapping functions that actually used to send/receive data and then decide whether the data should be tunneled (ie. if remote IP is within group/friendlist) or direct communication (ie. external/WAN IP such as official server), and we can make several mode of tunneling just like what i mentioned here http://forums.ppsspp.org/showthread.php?tid=16972&pid=121429#pid121429

ANR2ME commented on Core/HLE/sceNetAdhoc.cpp:
```diff
@@ -1637,7 +1640,7 @@ static int sceNetAdhocPtpOpen(const char *srcmac, int sport, const char *dstmac,
 			
 			// Valid Ports
-			if (!isPTPPortInUse(sport) && dport != 0) {
+			if (!isPTPPortInUse(sport) /*&& dport != 0*/) {
```
need to add 1 more line under this to fix connection issue with Phantasy Star Portable 2 when using non-zero PortOffset:

if (sport == 0) sport = -(int)portOffset; //making sure sport is 0 when shifted

adenovan commented:
that one more line used to fix phantasy star portable 2 using offset still have an issues.
its work only if windows host the party. 

if android host the party the game still throw network error occured  adb logcat show android can't connect to windows with error 114. 

joining to windows host is fine for android.

we still can't play phatasy star portable 2 betwen android device. 

if hosting the quest does the game bind to port 0 which is restricted to android?

influxWishdream commented:
This doesn't work because Android does not generally does not allow Port
Offset 0. Same with Linux. Though, port 0 isn't actually used in
communication but to randomize ports for communication instead.

I am not sure if Android has a firewall or something  that prevents Windows
to connect with Android.

Also, be sure that your port offset is greater than 1024. Anything under
that is generally occupied by important apps and the Android OS

On Nov 15, 2016 4:09 PM, "Ade Novan" <notifications@github.com> wrote:

> that one more line used to fix phantasy star portable 2 using offset still
> have an issues.
> its work only if windows host the party.
>
> if android host the party the game still throw network error occured adb
> logcat show android can't connect to windows with error 114.
>
> joining to windows host is fine for android.
>
> we still can't play phatasy star portable 2 betwen android device.
>
> if hosting the quest does the game bind to port 0 which is restricted to
> android?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hrydgard/ppsspp/pull/8134#issuecomment-260574679>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEcd5xVu9yFIRi8c6vhObwYPgibgc6Auks5q-WikgaJpZM4GZrZ0>
> .
>

ANR2ME commented:
Additionally to @influxWishdream answer, if i'm not mistaken socket error 114 = Operation already in progress (the socket in the middle of connecting but another attempt to connect is initiated), it's happened a lot (even on windows) on many games that uses non-blocking socket, it shouldn't be a problem because the games will repeatedly tried to connect until it eventually connected when using non-blocking socket.

adenovan commented:
@influxWishdream the offset value i was used is 1100 and its working on matching event but it can't enter the game lobby when android form party (hosting). the offset do the job like it should. its more like socket problem on linux that not found the equivalent error on windows.

@ANR2ME ah this is actually what im looking for right now. this 114 error problem is fixed by #9058 on monster hunter game which have same issues on matching peer. but its not work with this game. i tried to use port offset 0 and i see bind failed on adb logcat it's even worse and can't enter the matching event. so i think its not a port offset problem anymore and looks like something blocking the connection from android to windows when android form the party (firewall turned off and do it on lan and happened between android device too). if android join the party its works. is there something different between build when do PTPListen? looks like its not latency issues too. 

the weird part is windows log say its connected when joining to android as host. but android throw many socket error 114 even after executing ptpaccept, ptpsend ,and ptprecv which was respond to estabilished windows connection. windows successfuly connect to android and i think android just doesn't know that windows is connected. 

it seems android not entering this if condition  https://github.com/hrydgard/ppsspp/blob/master/Core/HLE/sceNetAdhoc.cpp#L1974
when form the party (hosting) and from the windows log i see the log its enter 3 times when hosting and says already connected 3 times after that says connection estabilished. tried to uncomment EALREADY on that if condition and both build can't enter the game after matching event it's even worse.

here is the logcat from android device and the difference when joining and hosting

hosting (failed)
[androidformdebug.txt](https://github.com/hrydgard/ppsspp/files/594698/androidformdebug.txt)

joining (success if windows host)
[windowsform.txt](https://github.com/hrydgard/ppsspp/files/594699/windowsform.txt)

ANR2ME commented:
Based on your androidformdebug.txt it seems it tried to connect to it self on listened port 13009 (instead of connecting to remote player/ppsspp) and didn't get accepted.

Are you sure port 13009 not taken by other program or being blocked?
You may want to live debug the source code and trace each line during sceNetAdhocPtpListen(OS¦Á,13009,8224,200000,300,5,0) to see whether it really successful in binding port 13009.

PS: as i remember adhoc functions is enabling Reusable sockets which shares the same port on multiple sockets, which probably a bad idea since already taken port can still be binded by another socket successfully, but the one accepting/receiving the data might be the other socket (which can be on a different program) and resulting a miscommunication.

adenovan commented:
im not sure its being taken or not because its happens on every android device, using netstat on my rooted device looks like no program bind to that port. also different offset value is not helping too. can we shift that ptp listen port using offset? i tried to do it but the client can't connect to the host.
 
![image](https://cloud.githubusercontent.com/assets/5246143/20384818/189ecf12-acf0-11e6-93c5-307cf8ac2445.png)

the log above disabled every SO_REUSEADDR on all scenetAdhoc source code and didn't help the game or give info socket error already used. sometimes its cause the game hang when disable it. 

left part off the image is android log listening (failed)
right part is windows log listening (success)

like you said its fails to connect to itself on port 13009 . didn't know if this issues can occured on linux build or not. someone need to test it and i think its easier to debug this issues using linux than android.

ANR2ME commented:
It's strange that it can accept connection from remote player/ppsspp on port 12000 but can't accept connection from it self on port 13009. 

Could it be the one listening and the one connecting have a different network interface/IP? 
you may want to print additional log to show the IP linked to the socket after it successfully binded and after it calls connect, to make sure both of them are using the same network interface/IP.

adenovan commented:
yeah great you're right shot into the problem straightly 👍 .

the game use different network interface /ip but this behaviour is not a problem on windows @_@. look at this log its connect to different network interface. 
![image](https://cloud.githubusercontent.com/assets/5246143/20400847/570f3d18-ad31-11e6-94dc-c860a4db89fd.png)

it should use 192.168.1.6 for local ip but its used my openvpn ip /interface. 

i can't log that INADDR_ANY properly after succesfully binded so i just add additional log on ptpconnect.

have any workaround for this issues? 

will post adb logcat soon..

adenovan commented:
that test use builtin adhoc server. the standalone and builtin adhoc server seems has the same issues. 
ok will do it look for the log on the server and trace the source also.

looks like its make sense the 192.168.12.1 is given by adhoc server. 
because this game connect to the server 2 times. first on matching event and after that do a connect again. can it be an ipconflict because its do a login after login and send 2 different ip to the server? AFAIK same ip is rejected from the server.

ANR2ME commented:
LOL didn't realize there is a constant IP there O.o
Anyway, gethostbyname does exist on linux but obsolete https://linux.die.net/man/3/gethostbyname

> The gethostbyname*() and gethostbyaddr*() functions are obsolete. Applications should use getaddrinfo(3) and getnameinfo(3) instead.

adenovan commented:
can we use this for get the addr info? looks like its work on macos too http://stackoverflow.com/questions/212528/get-the-ip-address-of-the-machine

hrydgard commented:
Okay, this needs a rebase now. And should we keep it around, or maybe merge it eventually?

hrydgard commented:
@ANR2ME  are you interested in bringing this up to date?

ANR2ME commented:
> 
> 
> @ANR2ME are you interested in bringing this up to date?

I think most of these already implemented by others in their forks, not sure whether they've been merged to the official one or not (i haven't touched PPSSPP for a long time)

This one should be closed i think, as this was intended for learning purpose only.
